dia-batching-server/src/api/polygon.rs

```diff
@@ -0,0 +1,393 @@
+use crate::api::error::PolygonError;
+use crate::api::Quotation;
+use crate::args::PolygonConfig;
+use crate::AssetSpecifier;
+use rust_decimal::Decimal;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+pub struct PolygonPriceApi {
+	client: PolygonClient,
+}
+
+impl PolygonPriceApi {
+	pub fn new_from_config(config: PolygonConfig) -> Self {
+		let api_key = config.pg_api_key.expect("Please provide a Polygon API key");
+
+		Self::new(config.pg_host_url, api_key)
+	}
+
+	pub fn new(host_url: String, api_key: String) -> Self {
+		let client = PolygonClient::new(host_url, api_key);
+
+		Self { client }
+	}
+
+	pub async fn get_prices(
+		&self,
+		assets: Vec<&AssetSpecifier>,
+	) -> Result<Vec<Quotation>, PolygonError> {
+		// Map used for the reverse lookup of ticker to asset specifier
+		let mut ticker_to_asset_map: HashMap<String, AssetSpecifier> = HashMap::new();
+
+		let from_currency_ticker_names = assets
+			.into_iter()
+			.filter_map(|asset| match Self::extract_source_currency(asset) {
+				Some(currency) => {
+					let source_currency_ticker =
+						PolygonClient::currency_to_ticker(currency.as_str());
+					// Insert the asset into the map
+					ticker_to_asset_map.insert(source_currency_ticker.clone(), asset.clone());
+					Some(source_currency_ticker)
+				},
+				None => {
+					log::warn!("Unsupported polygon asset: {:?}", asset);
+					None
+				},
+			})
+			.collect::<Vec<_>>();
+
+		let quotes = self.client.all_tickers(&from_currency_ticker_names).await?;
+
+		let mut prices = Vec::new();
+
+		// Add extra handling for USD if it was requested as it will not be in the quotes
+		if from_currency_ticker_names.contains(&PolygonClient::currency_to_ticker("USD")) {
+			let quotation = Quotation {
+				symbol: "USD-USD".to_string(),
+				name: "USD-USD".to_string(),
+				blockchain: Some("FIAT".to_string()),
+				price: Decimal::from(1),
+				supply: Decimal::from(0),
+				time: chrono::Utc::now().timestamp().unsigned_abs(),
+			};
+			prices.push(quotation);
+		}
+
+		for (ticker_name, quote) in quotes {
+			if let Some(asset) = ticker_to_asset_map.get(ticker_name.as_str()) {
+				let symbol = asset.symbol.clone();
+				// We use the bid price as the price
+				let price = quote.b;
```

Comment:
Do you know what is the price that the previous endpoint `conversions` was using? Because the bid price can sometimes be quite lower from the actual previous trade price.

Reply:
I compared the responses from the 'conversions' and the 'all-tickers' and it seemed to me like 'conversions' returned the bid price but rounded. 

See 
https://api.polygon.io/v1/conversion/AUD/USD?amount=1&precision=4&apiKey={API_KEY}
vs
https://api.polygon.io/v2/snapshot/locale/global/markets/forex/tickers?tickers=C:AUDUSD&apiKey={API_KEY}